sockeye/data_io.py

```diff
@@ -322,60 +322,72 @@ def shard_data(source_fname: str,
     """
     Assign int-coded source/target sentence pairs to shards at random.
 
-    :param source_fname: The file name of the source file.
+    :param source_fnames: The path to the source text (and optional token-parallel factor files).
     :param target_fname: The file name of the target file.
-    :param vocab_source: Source vocabulary.
-    :param vocab_target: Target vocabulary.
+    :param source_vocabs: Source vocabulary (and optional source factor vocabularies).
+    :param target_vocab: Target vocabulary.
     :param num_shards: The total number of shards.
     :param buckets: Bucket list.
     :param length_ratio_mean: Mean length ratio.
     :param length_ratio_std: Standard deviation of length ratios.
     :param output_prefix: The prefix under which the shard files will be created.
-    :return: Tuple of source, target file names and statistics for each shard as well as global statistics.
+    :return: Tuple of source (and source factor) file names, target file names and statistics for each shard,
+             as well as global statistics.
     """
+    source_fname, *source_factor_fnames = source_fnames
+    source_vocab, *source_factor_vocabs = source_vocabs
```

Comment:
Why separate them here, where in other parts of the code are handled as a single list? They are even put together several times in this same function, e.g. lines 367, 389. Line 381 could be combined with the following for loop. If it is only because of the different name, I think it would be ok to rename the prepared files to include .0. for the actual words (incompatible change then).

Reply:
I put them together to minimize the number of arguments to be passed around and make the concept clear that source surface form is the first factor.
I separate them in some files to make the changes backwards-compatible which I think is useful in this case.

Reply:
I agree that this looks a bit clumsy but the input/output signatures for each the functions is consistent.

Reply:
I actually like the fact that it is made clear that the first element is special and we maybe even argue that below where you have `sources[0]` it could be made clearer by doing the same kind of splitting instead.

Reply:
I actually generalized all the data_io functions to take multiple sources now which makes the whole unpacking unnecessary and much cleaner.